### PR TITLE
passed value decoder to interface Companion

### LINKED ISSUE #15125: interface view decoders
Build upon #14313, and support #14969, by having `InterfaceCompanion` also supply a `ValueDecoder<View>`.

* [x] #15017
* [x] `InterfaceCompanion`'s tparams should be `<I, View>`
* [x] `ContractTypeCompanion`'s tparams should be `<Marker, Data>`. `ContractCompanion` should pass `Data` for *both*. They should be `I, View` for `InterfaceCompanion`
* [x] change the `INTERFACE` class we generate to pass the view type as the second tparam, and pass a `ValueDecoder<View>` as a param to `super`

### PR DESCRIPTION
**Pull request by chunlokling-da**
fixes #15125

CHANGELOG_BEGIN
CHANGELOG_END

### CODE REVIEW

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala [hunk 1]
```diff
@@ -46,11 +46,15 @@ object ClassForType extends StrictLogging {
     for {
       (interfaceName, interface) <- typeWithContext.interface.interfaces.toList
       className = ClassName.bestGuess(fullyQualifiedName(interfaceName))
+      interfaceViewTypeName = ClassName.bestGuess(
+        fullyQualifiedName(interface.viewType.get, packagePrefixes)
```
Comment (chunlokling-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala:50]: viewType can be optional. i will fix it

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala [hunk 2]
```diff
@@ -46,11 +46,22 @@ object ClassForType extends StrictLogging {
     for {
       (interfaceName, interface) <- typeWithContext.interface.interfaces.toList
       className = ClassName.bestGuess(fullyQualifiedName(interfaceName))
+      interfaceViewTypeName = ClassName.bestGuess(
```
Comment (chunlokling-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala:49]: Is it correct to assume view type is required for all interfaces?

File: language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala [hunk 3]
```diff
@@ -46,11 +46,22 @@ object ClassForType extends StrictLogging {
...
+        fullyQualifiedName(
+          interface.viewType.getOrElse(
+            throw new IllegalArgumentException(
+              s"View Type is required for interface $interfaceName"
+            )
```
Comment (chunlokling-da) [inline language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala:54]: Is it correct to assume view type is required for all interfaces?

Comment (S11001001) [reply language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/backend/java/inner/ClassForType.scala:54]: It is correct. Moreover, it is always a type `T` that `extends codegen.DamlRecord<T>`, if you have a reason to exploit that.

File: language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractId.java [hunk 4]
```diff
@@ -54,7 +54,7 @@ public final ExerciseCommand makeExerciseCmd(String choice, Value choiceArgument
 
   // overridden by every code generator, but decoding abstractly can e.g.
   // produce a ContractId<Foo> that is not a Foo.ContractId
-  protected ContractTypeCompanion getCompanion() {
+  protected ContractTypeCompanion<T, T> getCompanion() {
```
Comment (S11001001) [inline language-support/java/bindings/src/main/java/com/daml/ledger/javaapi/data/codegen/ContractId.java:57]:
```suggestion
  protected ContractTypeCompanion<T, ?> getCompanion() {
```

`T` in the second position is wrong for interfaces.

**Review by ray-roestenburg-da — APPROVED** [review]

**Pull request merged**